addons/stock/models/stock_picking.py

```diff
@@ -583,8 +583,13 @@ def action_assign(self):
 
     @api.multi
     def action_cancel(self):
-        self.mapped('move_lines')._action_cancel()
-        self.write({'is_locked': True})
+        for picking in self:
+            value = {'is_locked': True}
+            if picking.move_lines:
+                picking.move_lines._action_cancel()
+            else:
+                value['state'] = 'cancel'
+            picking.write(value)
```

Comment:
Maybe we can keep it more batchable

``` python
picking_with_moves = self.env['stock.picking']
picking_without_moves = self.env['stock.picking']
for picking in self:
    if picking.move_lines:
        picking_with_moves |= picking
    else:
        picking_without_moves |= picking

# If the picking is linked to stock moves, cancel them and
# lock the picking.
picking_with_moves.move_lines._action_cancel()
picking_with_moves.write({
    'is_locked': True,
})

# If the picking is not linked to stock moves, for the state
# to cancel and lock the picking.
picking_without_moves.write({
    'is_locked': True,
    'state': 'cancel',
})
return True
```

and you need to add two tests in test_move2.py in testsinglepicking

test_cancel_1 > Cancel a confirmed picking with a stock move

test_cancel_2 > Cancel a confirmed picking with no stock move but a stock move line
in both test, assert the picking is confirmed, call cancel, assert the picking is cancel
in test_cancel_2, make sure to set the immediate transfert flag to True

Reply:
in case of imediate transfer, the state is ( **"Draft"** )  so we can build test case with the manual transfer.

Reply:
I would rather try to remove
https://github.com/odoo/odoo/blob/master/addons/stock/models/stock_move_line.py#L171
https://github.com/odoo/odoo/blob/master/addons/stock/models/stock_picking.py#L625 
Make the move creation become a normal behavior (the initial demand on move and the state would require some modification in the stock.move.line.py file).
If the move is in state done just bypass the action_confirm